internal/services/oidc.go

```diff
@@ -0,0 +1,250 @@
+		msg := "An error occurred when trying to receive the session <nil>"
+		osvc.logger.Error(msg)
+		http.Error(w, msg, http.StatusBadRequest)
+		return nil
+	}
+
+	nonceToken, err := osvc.GetValueFromToken("nonce", token)
+	if err != nil {
+		msg := fmt.Sprintf("An unexpected error occurred when getting the nonce from ID Token. Err: %v", err)
+		osvc.logger.Errorf(msg)
+		http.Error(w, msg, http.StatusBadRequest)
+		return nil
+	}
+	if !session.MatchNonce(nonceToken.(string), r) {
+		msg := "nonce did not match"
+		osvc.logger.Error(msg)
+		http.Error(w, msg, http.StatusBadRequest)
+		return nil
+	}
+
+	jwt, err := osvc.createJWT(&token.CustomClaims)
+	if err != nil {
+		msg := fmt.Sprintf("An error occurred while creating a JWT. Error: %v", err)
+		osvc.logger.Error(msg)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return nil
+	}
+
+	// If all goes well, then create the JSON data to display as body content.
+	jsonBody, err := json.Marshal(jwt)
+	if err != nil {
+		msg := fmt.Sprintf("An error occurred while validating some tokens. Error: %v", err)
+		osvc.logger.Errorf(msg)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return nil
+	}
+
+	return jsonBody
+}
+
+// claimsToken creates token claims.
+func (osvc *OIDCService) claimsToken(ctx context.Context,
+	w http.ResponseWriter, r *http.Request) (*models.Tokens, error) {
+
+	oauth2Token, err := osvc.createOAuth2Token(ctx, w, r)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to exchange token. Error: %v", err)
+		osvc.logger.Error(msg)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return nil, err
+	}
+	idToken, err := osvc.validateIDToken(ctx, oauth2Token, w, r)
+	if err != nil {
+		msg := fmt.Sprintf("An error occurred while validating ID Token. Error: %v", err)
+		osvc.logger.Error(msg)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return nil, err
+	}
+
+	idTokenClaims := new(json.RawMessage)
+	if err := idToken.Claims(&idTokenClaims); err != nil {
+		msg := fmt.Sprintf("An error occurred while validating ID Token. Error: %v", err)
+		osvc.logger.Error(msg)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return nil, err
+	}
+
+	userInfo, err := osvc.GetUserInfo(ctx, oauth2Token)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to get UserInfo. Error: %v", err)
+		osvc.logger.Error(msg)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return nil, err
+	}
+
+	cc := models.CustomClaims{}
+	if err := idToken.Claims(&cc); err != nil {
+		msg := fmt.Sprintf("An unexpected error has occurred. Error: %v", err)
+		osvc.logger.Error(msg)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return nil, err
+	}
+
+	return &models.Tokens{
+		OAuth2Token:   oauth2Token,
+		IDTokenClaims: idTokenClaims,
+		UserInfo:      userInfo,
+		CustomClaims:  cc,
+	}, nil
+}
+
+// createOAuth2Token creates a new OAuth2 token.
+func (osvc *OIDCService) createOAuth2Token(ctx context.Context,
+	w http.ResponseWriter, r *http.Request) (*oauth2.Token, error) {
```

Comment:
Pass the authorization `code` as a parameter instead of `w http.ResponseWriter`, `r *http.Request`.